// @mistralai/mistralai/models/components/fileschema.js
"use strict";
/*
 * Code generated by Speakeasy (https://speakeasy.com). DO NOT EDIT.
 */
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.FileSchema$ = exports.FileSchema$outboundSchema = exports.FileSchema$inboundSchema = exports.FileSchemaPurpose$ = exports.FileSchemaPurpose$outboundSchema = exports.FileSchemaPurpose$inboundSchema = exports.FileSchemaPurpose = void 0;
const z = __importStar(require("zod"));
const primitives_js_1 = require("../../lib/primitives.js");
const enums_js_1 = require("../../types/enums.js");
const sampletype_js_1 = require("./sampletype.js");
const source_js_1 = require("./source.js");
/**
 * The intended purpose of the uploaded file. Only accepts fine-tuning (`fine-tune`) for now.
 */
exports.FileSchemaPurpose = {
    FineTune: "fine-tune",
};
/** @internal */
exports.FileSchemaPurpose$inboundSchema = z
    .union([
    z.nativeEnum(exports.FileSchemaPurpose),
    z.string().transform(enums_js_1.catchUnrecognizedEnum),
]);
/** @internal */
exports.FileSchemaPurpose$outboundSchema = z.union([
    z.nativeEnum(exports.FileSchemaPurpose),
    z.string().and(z.custom()),
]);
/**
 * @internal
 * @deprecated This namespace will be removed in future versions. Use schemas and types that are exported directly from this module.
 */
var FileSchemaPurpose$;
(function (FileSchemaPurpose$) {
    /** @deprecated use `FileSchemaPurpose$inboundSchema` instead. */
    FileSchemaPurpose$.inboundSchema = exports.FileSchemaPurpose$inboundSchema;
    /** @deprecated use `FileSchemaPurpose$outboundSchema` instead. */
    FileSchemaPurpose$.outboundSchema = exports.FileSchemaPurpose$outboundSchema;
})(FileSchemaPurpose$ || (exports.FileSchemaPurpose$ = FileSchemaPurpose$ = {}));
/** @internal */
exports.FileSchema$inboundSchema = z.object({
    id: z.string(),
    object: z.string(),
    bytes: z.number().int(),
    created_at: z.number().int(),
    filename: z.string(),
    purpose: z.literal("fine-tune").optional(),
    sample_type: sampletype_js_1.SampleType$inboundSchema,
    num_lines: z.nullable(z.number().int()).optional(),
    source: source_js_1.Source$inboundSchema,
}).transform((v) => {
    return (0, primitives_js_1.remap)(v, {
        "created_at": "createdAt",
        "sample_type": "sampleType",
        "num_lines": "numLines",
    });
});
/** @internal */
exports.FileSchema$outboundSchema = z.object({
    id: z.string(),
    object: z.string(),
    bytes: z.number().int(),
    createdAt: z.number().int(),
    filename: z.string(),
    purpose: z.literal("fine-tune").default("fine-tune"),
    sampleType: sampletype_js_1.SampleType$outboundSchema,
    numLines: z.nullable(z.number().int()).optional(),
    source: source_js_1.Source$outboundSchema,
}).transform((v) => {
    return (0, primitives_js_1.remap)(v, {
        createdAt: "created_at",
        sampleType: "sample_type",
        numLines: "num_lines",
    });
});
/**
 * @internal
 * @deprecated This namespace will be removed in future versions. Use schemas and types that are exported directly from this module.
 */
var FileSchema$;
(function (FileSchema$) {
    /** @deprecated use `FileSchema$inboundSchema` instead. */
    FileSchema$.inboundSchema = exports.FileSchema$inboundSchema;
    /** @deprecated use `FileSchema$outboundSchema` instead. */
    FileSchema$.outboundSchema = exports.FileSchema$outboundSchema;
})(FileSchema$ || (exports.FileSchema$ = FileSchema$ = {}));
